pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -323,10 +325,11 @@ private void grantPermissions(String topicUri, String role, Set<AuthAction> acti
         try {
             namespaceResources().set(path(POLICIES, namespaceName.toString()), (policies) -> {
                 if (!policies.auth_policies.destination_auth.containsKey(topicUri)) {
-                    policies.auth_policies.destination_auth.put(topicUri, new HashMap<>());
+                    policies.auth_policies.destination_auth.put(topicUri, new ConcurrentHashMap<>());
                 }
-
-                policies.auth_policies.destination_auth.get(topicUri).put(role, actions);
+                Set<AuthAction> authActionSet = Collections.newSetFromMap(new ConcurrentHashMap<>());
+                authActionSet.addAll(actions);
```

Comment:
I'd recommend using the `Map.compute` method here instead of separate lookups and mutations since besides fixing "lost updates", the code is cleaner also when concurrency is not a concern.